crates/bevy_render/src/render_resource/resource_macros.rs

```diff
@@ -120,4 +120,29 @@ macro_rules! render_resource_wrapper {
     };
 }
 
+#[macro_export]
+macro_rules! define_atomic_id {
+    ($atomic_id_type:ident) => {
+        #[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
+        pub struct $atomic_id_type(u32);
+
+        impl Default for $atomic_id_type {
+            fn default() -> Self {
+                use std::sync::atomic::{AtomicU32, Ordering};
+
+                static COUNTER: AtomicU32 = AtomicU32::new(0);
+                COUNTER
+                    .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |val| {
```

Comment:
`AtomicU32::fetch_add` is probably more succinct and performant.

Reply:
The issue is that `AtomicU32::fetch_add` wraps around on overflow. This is undetectable and could theoretically lead to undefined behavior. Whether it is likely or not that reusing IDs will lead to failure is probably questionable, especially when we consider that UUIDs could potentially be the same as well.

With `fetch_update` in place, we can also reevaluate the u32 choice if people were to run into the limit.

Alternatively, we could set the initial id to 1 and panic if any ID ever reaches 0. This would save us from the synchronization overhead caused by the `fetch_update`.

Reply:
Either a comment justifying this or the proposed check would be great.

If all of the use cases can support it, we could also use the NonZeroU32 type to niche these IDs as well, but that might not be 100% necessary as it's not that commonly required.

Reply:
The check is in place. 👍 
Yeah, I think this solution is simple enough to understand. And if it turns out that the u32 limit is not big enough, then we can increase it at any time.